File: 2022/3/solution.py
import os
from typing import Dict, List, Set

ALPHABET = '_abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'
PRIORITIES: Dict[str, int] = {c: i for i, c in enumerate(ALPHABET)}

def calculate_priorities(*args):
  return sum([PRIORITIES[c] for c in set.intersection(*args)])

def part1(puzzle_input: List[str]) -> int:
  res = 0
  for line in puzzle_input:
    n = len(line)
    comp1: Set[str] = set(line[:n//2])
    comp2: Set[str] = set(line[n//2:])
    res += calculate_priorities(comp1, comp2)
  return res

def part2(puzzle_input: List[str]) -> int:
  res = 0
  for i in range(0, len(puzzle_input), 3):
    ruck1: Set[str] = set(puzzle_input[i])
    ruck2: Set[str] = set(puzzle_input[i + 1])
    ruck3: Set[str] = set(puzzle_input[i + 2])
    res += calculate_priorities(ruck1, ruck2, ruck3)
  return res

if __name__ == '__main__':
  with open(os.path.join(os.path.dirname(os.path.abspath(__file__)), 'input.txt'),'r') as file:
    puzzle_input = file.read().strip().split('\n')
  print(part1(puzzle_input))
  print(part2(puzzle_input))

File: 2022/3/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := strings.Split(string(data), "\n")
	fmt.Println(part1(puzzleInput))
	fmt.Println(part2(puzzleInput))
}

func makeSet(s string) map[rune]struct{} {
	set := make(map[rune]struct{}, len(s))
	for _, c := range s {
		set[c] = struct{}{}
	}
	return set
}

func calcPriority(r rune) int {
	if r > 'Z' {
		return int(r-'a') + 1
	}
	return int(r-'A') + 27
}

func part1(puzzleInput []string) int {
	res := 0
	for _, line := range puzzleInput {
		l := len(line)
		set1 := makeSet(line[:l/2])
		set2 := makeSet(line[l/2:])
		for c := range set1 {
			if _, exists := set2[c]; exists {
				res += calcPriority(c)
				break
			}
		}
	}
	return res
}

func part2(puzzleInput []string) int {
	res := 0
	for i := 0; i < len(puzzleInput); i += 3 {
		set1 := makeSet(puzzleInput[i])
		set2 := makeSet(puzzleInput[i+1])
		set3 := makeSet(puzzleInput[i+2])
		for c := range set1 {
			_, exists1 := set2[c]
			_, exists2 := set3[c]
			if exists1 && exists2 {
				res += calcPriority(c)
				break
			}
		}
	}
	return res
}
